### PR TITLE
Convert project to typescript

### PR DESCRIPTION
**Pull request by Nitwel**
First try on converting the code into Typescript.

- [ ] This PR  only introduces changes to documentation.
  - Please include a summary of changes and an explanation
- [ ] This PR adds new functionality
  - [ ] Is there a related issue?
    - Please note the related issue below, and how this PR relates to the issue if appropriate
  - [ ] Does the code style reasonably match the existing code?
  - [x] Are the changes tested (using the existing format, as far as is possible?)
  - [ ] Are the changes documented in the readme with a suitable example?
  - [ ] Is the table of contents updated?
- [x] This PR introduces some other kind of change
  - Updated the project to use typescript and decentralize code into many files.

### CODE REVIEW

**Review by francisrstokes — CHANGES_REQUESTED** [review]

**Comment by francisrstokes** [comment]
OK so overall there are 3 big hurdles to overcome with this PR:

1. Finding a solution to the TextEncoder/TextDecoder problem. It needs to work in both node and the browser, and it needs to be able to resolve this synchronously. You'll probably need to look into adding at least `@types/node` to the project - maybe doing some research to see if there is some kind of universal text module?
2. The build system needs to be able to compile the TS and bundle it up into a single file. That file should, ideally, target modern JS, and not introduce a lot of interop cruft. This will probably require playing around with the compiler options in the tsconfig.
3. The style and formatting of the project is a bit all over the place right now. I actually made a PR template with a checklist of items - and one of them was about matching project style. I've edited your PR to include the checklist - please use it as a guide. You don't have to check every box if it isn't relevant, but filling it in will help avoid nitpicky comments and generally makes it easier to review.

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,1091 @@
...
+      cp = cp >> 8;
+      i++;
+    }
+    total++;
+  }
+  return total;
+}
+
+//    createParserState :: x -> s -> ParserState e a s
+const createParserState = <D>(target: InputType, data: D | null = null): ParserState<null, string | null, D | null> => {
```
[the remark this answers is not in the document]
Comment (Nitwel) [reply index.ts]: Otherwhise TS is saying that D can't be instantiated as null

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,1091 @@
...
+    const p = this.p;
+    return new Parser(function Parser$map$state(state): ParserState<T | T2, E, D> {
+      const newState = p(state);
+      if (newState.isError) return newState;
+      return updateResult(newState, fn(newState.result));
+    });
+  };
+  
+  //                chain :: Parser e a s ~> (a -> Parser e b s) -> Parser e b s
+  chain<T2>(fn: (x?: T) => Parser<T2, E, D>): Parser<T | T2, E, D> {
```
[the remark this answers is not in the document]
Comment (Nitwel) [reply index.ts]: Yes and it makes sense that it could be T or T2 as we return T when the state is an error.

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      "moduleResolution": "Node",
+      "esModuleInterop": false,
+      "strict": true,
+      "noImplicitAny": true,
+      "sourceMap": true,
+      "outDir": "dist",
+      "baseUrl": ".",
+      "paths": {
+        "*": [
+          "node_modules/",
```
Comment (francisrstokes) [inline tsconfig.json:14]: Is this part necessary?

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (francisrstokes) [inline tsconfig.json:1]: I guess we also need to tell TS to generate a `index.d.ts` file for the repo as well (in the root directory)

**Comment by francisrstokes** [comment]
@Nitwel It's looking really good - awesome work! Just a couple of small points for the build and I think we're almost there:

- We should be building two versions of the library: one based on ESModules, and one based on commonjs. So we might need a couple of different configs for TS and rollup to make that work, but if there's simpler way I'm open to it
- The builds should be just live in the root directory as `index.mjs` for the ESModules build and `index.js` for the commonjs build. This matches previous releases and enables the `package.json` to point each user to the right JS file for their setup

I also mentioned a couple of things about the tsconfig in a separate comment.

File: .gitignore [hunk 5]
```diff
@@ -1,7 +1,6 @@
 node_modules/
 .sketchpad/
-/index.js
+/index.*
```
Comment (francisrstokes) [inline .gitignore:3]: This will also catch `index.d.ts` file 😄

Comment (francisrstokes) [reply .gitignore:3]: Ah I guess we don't need to commit it

Comment (Nitwel) [reply .gitignore:3]: yep, that's what I thought.

File: rollup.config.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import dts from "rollup-plugin-dts";
+
+export default [
+  {
+    input: './src/index.ts',
+    output: {
+        file: 'index.js',
+        format: 'cjs'
+    },
+    plugins: [typescript()]
```
Comment (francisrstokes) [inline rollup.config.js:11]: I noticed some oddities in the build files (check line 120 of index.js - it contains a `p;` statement which is of course not valid in JS). This is why the tests are failing for node 10. Apparently later versions can tolerate that 🤔 

In any case, we can specify the build version as ES6 for the common js build, and that takes care of it.

```javascript
plugins: [typescript({
  target: "ES6"
})]
```

File: rollup.config.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  },
+  {
+    input: './src/index.ts',
+    output: {
+        file: 'index.mjs',
+        format: 'es'
+    },
+    plugins: [typescript()]
```
Comment (francisrstokes) [inline rollup.config.js:19]: For the ESModules build let's target ES2020.

**Review by francisrstokes — CHANGES_REQUESTED** [review]

**Comment by francisrstokes** [comment]
I've published a release candidate for this PR (4.0.0-rc.0) which you can test out by installing `arcsecond@next`. I'm going to ask some people to try it out - please also test it out yourself. There might be some changes to be made in response, but I think the core of it is done and looking good. I'm always skeptical about taking a project and rewriting it in another language, but I have to say you've done a great job here @Nitwel and I think the results show. I'm super excited to _finally_ be getting proper TS support for arcsecond - and of course the bonus that comes from the types when still using arcsecond in JS.

**Review by francisrstokes — APPROVED** [review]

**Pull request merged**